sdk/core/Azure.Core/tests/TestFramework/RecordSession.cs

```diff
@@ -82,6 +82,11 @@ public void Sanitize(RecordedTestSanitizer sanitizer)
                 {
                     sanitizer.Sanitize(entry);
                 }
+
+                foreach (var variable in Variables.ToArray())
+                {
+                    Variables[variable.Key] = sanitizer.SanitizeVariable(variable.Key, variable.Value);
```

Comment:
What does this end up doing? Looking at https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/tests/TestFramework/RecordedTestSanitizer.cs#L42 it seems like it would always just give the value back.

Reply:
We override this with custom sanitizers per service.  See https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs#L11 for example.

Reply:
Don't we still risk leaking secrets if one of the test libraries don't correctly override this? Should we change the default to be empty or at least make it abstract so we don't end up accidently leaking something?

Reply:
Also with the new TestEnvironment work don't we know all the environment variables? Can we just blank them out for everyone?

Reply:
> Don't we still risk leaking secrets if one of the test libraries don't correctly override this? Should we change the default to be empty or at least make it abstract so we don't end up accidently leaking something?

Most libraries don't record secrets so we don't want to break them by default, mostly it's the ones that don't support AD and have to use connection strings and keys.

> Also with the new TestEnvironment work don't we know all the environment variables? Can we just blank them out for everyone?

Let me describe why we need to record these values at all.

Imagine you have a client that takes a connection string that contains host and key like this "Host=pakrym.com;Key=secret123". Then it uses the host to generate URIs to send requests to and key to format the authentication header.

It all works well when tests are running live and we are reading values from environment variables. Things start getting interesting when you need to record request/responses and play them back without having the environment variables set.

Our playback framework requires request URL and headers to match precisely what was recorded (except the sanitized headers like Authentication) so during the playback we need to give clients enough inputs so it generates the same URL as during live run.

This is why we record "Host=pakrym.com;Key=secret123" as part of test record session file and when tests are running as playback we read them back and pass to client constructors.

Naturally we can't store the secret in the file that would get pushed to github but it's not a problem because secret is used to generate the `Authentication` header which value we don't compare anyway. So we are able to sanitize real secret from connection string and replace it with a dummy value  "Host=pakrym.com;Key=Sanitized" (https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/appconfiguration/Azure.Data.AppConfiguration/tests/SessionRecords/ConfigurationLiveTests/AddSetting.json#L117). 

For services that use TokenCredential we don't have to record any secret values and we only end up storing the endpoint used so requests can be recreated the same way. 
(https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Keys/tests/SessionRecords/KeyClientLiveTests/CreateKey.json#L248)

Reply:
Thanks for the explanation but I do still worry a little about potentially missing a secret that we should sanitize, not sure if there is any good way to guard against it though.

Reply:
One thing we can do is a whitelist of connection string parameters. By default we would sanitize all values in anything that looks like connection string and you have to register parameter names are allowed. Won't work with non-connection string variables though.